frontend/lib/resolution/InitResolver.cpp

```diff
@@ -313,6 +313,38 @@ static const Type* ctFromSubs(Context* context,
     auto manager = AnyOwnedType::get(context);
     auto dec = ClassTypeDecorator(ClassTypeDecorator::BORROWED_NONNIL);
     ret = ClassType::get(context, basic, manager, dec);
+  } else if (receiverType->isDomainType()) {
+    // Extract domain information from _instance substitution
+    // TODO: also support associative domains here
+
+    QualifiedType instanceQt;
+    QualifiedType rank;
+    QualifiedType idxType;
+    QualifiedType stridable;
+
+    if (subs.size() == 1) {
+      instanceQt = subs.begin()->second;
+      if (auto instance = instanceQt.type()) {
+        if (auto instanceCt = instance->toClassType()) {
+          if (auto instanceBct = instanceCt->basicClassType()) {
+            // Get BaseRectangularDom parent subs for rectangular domain info
+            if (auto baseRect = instanceBct->parentClassType()) {
+              if (baseRect->name() == "BaseRectangularDom") {
+                auto innerSubs = baseRect->sortedSubstitutions();
+                CHPL_ASSERT(innerSubs.size() == 3);
+
+                rank = innerSubs[0].second;
+                idxType = innerSubs[1].second;
+                stridable = innerSubs[2].second;
+              }
+            }
+          }
+        }
+      }
+    }
```

Comment:
I do think it's a lot of levels of nesting and it should go in a helper.
But to clarify, which conditionals do you mean I could invert? I think the only one that would make sense is turning the outermost into a guard that returns early, since the others are also responsible for setting variables. Or would you rather have them all be a series of un-nested conditionals that at each step check if we want to return early?

Reply:
That's a good point. I think a flat version of this code which intersperses variable declarations with inverted conditionals that return early would look nicer, though it would be a tad longer since you can't combine variable declarations with ifs anymore. But I'm not the style police -- if you feel that it looks better as is, feel free to leave it.